drivers/dht/dht.c

```diff
@@ -17,76 +18,94 @@
  *
  * @author      Ludwig Knüpfer <ludwig.knuepfer@fu-berlin.de>
  * @author      Christian Mehlis <mehlis@inf.fu-berlin.de>
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
  *
  * @}
  */
 
 #include <stdint.h>
+#include <string.h>
 
+#include "log.h"
 #include "xtimer.h"
 #include "timex.h"
 #include "periph/gpio.h"
 
 #include "dht.h"
+#include "dht_params.h"
 
 #define ENABLE_DEBUG    (0)
 #include "debug.h"
 
-/***********************************************************************
- * internal API declaration
- **********************************************************************/
+#define PULSE_WIDTH_THRESHOLD       (40U)
 
-static void dht_read_data(gpio_t dev, uint32_t *data, uint8_t *checksum);
-static int dht_test_checksum(uint32_t data, uint8_t checksum);
-void dht_parse_11(dht_data_t *data, float *outhum, float *outtemp);
-void dht_parse_22(dht_data_t *data, float *outhum, float *outtemp);
-
-/***********************************************************************
- * internal API implementation
- **********************************************************************/
+/**
+ * @brief   Allocation of memory for device descriptors
+ */
+dht_t dht_devs[DHT_NUMOF];
 
-void dht_parse_11(dht_data_t *data, float *outhum, float *outtemp)
+static uint16_t read(gpio_t pin, int bits)
 {
-    *outhum = data->humidity >> 8;
-    *outtemp = data->temperature >> 8;
+    uint32_t start, end;
+    uint16_t res = 0;
+
+    for (int i = 0; i < bits; i++) {
+        res <<= 1;
+        /* measure the length between the next rising and falling flanks (the
+         * time the pin is high - smoke up :-) */
+        while (!gpio_read(pin));
+        start = xtimer_now();
+        while (gpio_read(pin));
+        end = xtimer_now();
+        /* if the high phase was more than 40us, we got a 1 */
+        if ((end - start) > PULSE_WIDTH_THRESHOLD) {
+            res |= 0x0001;
```

Comment:
no need for it: a substraction between two unsigned ints is overflow agnostic... Example:
```
uint8_t a = 250, b = 240, c = 5;
a - b := 10;
c - a := 10;
```
so there is no need whatsoever to manually catch any overflow...